crates/node/src/keyshare.rs

```diff
@@ -313,15 +326,47 @@ impl KeyshareStorage {
             anyhow::bail!("backup keyshares is not an exact match for the contract keyset")
         }
 
-        // Ensure we import into an empty Keystore
-        let permanent = self.permanent.load().await?;
-        if permanent.is_some() {
-            anyhow::bail!("permanent keyshare storage isn't empty");
+        // We load all keys we have stored in permanent or temporary storage and only draw from
+        // backup in case we are missing shares.
+        let existing_keyshares = self._load_prefix_from_permanent(contract_keyset).await?;
+        let mut new_keyshares = existing_keyshares;
+        for domain in contract_keyset.domains.iter().skip(new_keyshares.len()) {
+            let key_id =
+                KeyEventId::new(contract_keyset.epoch_id, domain.domain_id, domain.attempt);
+
+            // if we don't have the keyshare in temporary storage, we load it from the backup
+            let keyshare: Keyshare = self
+                .temporary
+                .load_keyshare(key_id)
+                .await?
+                .or_else(|| backup.iter().find(|share| share.key_id == key_id).cloned())
+                .ok_or_else(|| anyhow::anyhow!("Missing keyshare {:?}", key_id))?;
```

Comment:
Nit:

> Error messages are typically concise lowercase sentences without trailing punctuation

From https://doc.rust-lang.org/std/error/trait.Error.html

```suggestion
                .ok_or_else(|| anyhow::anyhow!("missing keyshare {:?}", key_id))?;
```